containers/sip-servlets-as7/src/main/java/org/jboss/as/clustering/web/infinispan/sip/DistributedCacheManager.java

```diff
@@ -0,0 +1,1302 @@
+		if (logger.isDebugEnabled()){
+			logger.debug("storeSipSessionData - sipSessionKey=" + sipSessionKey + "--> cache key");
+			logger.debug("storeSipSessionData - sipSessionId=" + sipSessionId);
+			logger.debug("storeSipSessionData - sipApplicationSessionKey=" + sipApplicationSessionKey);
+		}
+		
+		Operation<Void> operation = new Operation<Void>() {
+            @Override
+            public Void invoke(Cache<String, Map<Object, Object>> cache) {
+            	Map<Object, Object> map = cache.putIfAbsent(sipSessionKey, null);
+            	try {
+            		
+	                SipSessionMapEntry.VERSION.put(map, Integer.valueOf(sessionData.getVersion()));
+	                DistributableSipSessionMetadata dsm = (DistributableSipSessionMetadata)(sessionData.getMetadata());
+	                
+	                if (logger.isDebugEnabled()){
+            			logger.debug("storeSipSessionData - dsm=" + dsm);
+            			if (dsm != null){
+            				logger.debug("storeSipSessionData - dsm.isNew()=" + dsm.isNew());
+            				logger.debug("storeSipSessionData - dsm.isValid()=" + dsm.isValid());
+            				logger.debug("storeSipSessionData - sessionData.isSessionMetaDataDirty()=" + sessionData.isSessionMetaDataDirty());
+            			}
+            		}
+	                
+	                if (dsm != null && sessionData.isSessionMetaDataDirty()){
+		                if (dsm.isNew()){
+		                	
+		                	if (sessionData.getMetadata().getId() == null){
+		                		// Setting the id manually. Otherwise it would cause exceptions.
+		                		if (logger.isDebugEnabled()){
+		                			logger.debug("storeSipSessionData - sessionData.getMetadata().getId() is null - setting id manually to " + sipSessionId);
+		                		}
+		                		sessionData.getMetadata().setId(sipSessionId);
+		                	}
+		                	SipSessionMapEntry.METADATA.put(map, sessionData.getMetadata());
+		                }
+		                
+		                if (dsm.getMetaData() != null){
+		                	if (logger.isDebugEnabled()){
+		            			logger.debug("storeSipSessionData - storing SIP_METADATA_MAP: " + dsm.getMetaData());
+		            			for (String tmpKey: dsm.getMetaData().keySet()){
+		            				logger.debug("storeSipSessionData - storing SIP_METADATA_MAP, entry: " + tmpKey + "=" + dsm.getMetaData().get(tmpKey));	
+		            			}
+		            		}
+		                	SipSessionMapEntry.SIP_METADATA_MAP.put(map, DistributedCacheManager.this.marshaller.marshal(dsm.getMetaDataCopy()));
+		                }
+		                
+	                }   
+	                if (sessionData.getTimestamp() != null){
+	                	if (logger.isDebugEnabled()){
+	                		logger.debug("storeSipSessionData - SipSessionMapEntry.TIMESTAMP.put");
+	                	}
+	                	SipSessionMapEntry.TIMESTAMP.put(map, sessionData.getTimestamp());
```

Comment:
This is also the same way as how web replication works in JBoss codes. See class org.jboss.as.clustering.web.infinispan.DistributedCacheManager. (Anyway, there is only one map instance which gets updated multiple times as VERSION,METADATA,TIMESTAMP are put into it.)